### PR TITLE
allow availability recheck for external storages

### LINKED ISSUE #22790: External storages will be marked unavailable until manually checked
With #20920 (specifically 2e3a5ccaaa1a64617980cd086a3dc13a9058892b) code was added such that any external storages that are marked as unavailable will be fake-mounted using FailedStorage, instead of the real storage. This was probably done so that the red row highlight works properly, which will trigger a storage recheck when the root files view of the web UI is loaded asynchronously through `StorageService::updateStorageStatus()`. However, that code will not run if the storage is only accessed through the DAV interface, or even if the root files view is never loaded (I think, not entirely sure on that point).

When a storage is marked as unavailable, it should still be mounted into the filesystem, so that the filesystem can decide to recheck the availability if the recheck TTL has expired. A storage should never be prevented from being mounted based on its availability.

cc @icewind1991 @PVince81

### PR DESCRIPTION
**Pull request by icewind1991**
Fixes #22790

cc @Xenopathic @DeepDiver1975 @PVince81

### CODE REVIEW

File: apps/files_external/lib/config/configadapter.php [hunk 1]
```diff
@@ -132,7 +133,7 @@ public function getMountsForUser(IUser $user, IStorageFactory $loader) {
 
 			try {
 				$availability = $impl->getAvailability();
-				if (!$availability['available']) {
+				if (!$availability['available'] && !Availability::shouldRecheck($availability)) {
 					$impl = new FailedStorage(['exception' => null]);
```
Comment (PVince81) [inline apps/files_external/lib/config/configadapter.php]: Can you also fix the null exception problem too here ? You can cherry pick from my PR if you like.

Comment (icewind1991) [inline apps/files_external/lib/config/configadapter.php]: done

**Comment by PVince81** [comment]
Thanks, the code looks good. Will test in a few.

**Comment by Xenopathic** [comment]
Why are we mounting with a FailedStorage at all? Just remove the whole bit that checks the availability in files_external, let the filesystem handle it.

:-1:

**Comment by icewind1991** [comment]
> Why are we mounting with a FailedStorage at all?

Because if we don't the mounts will be filtered out due to their root valid, without that logic failed storages don't show up in the ui

**Comment by Xenopathic** [comment]
@icewind1991 I thought the logic was supposed to be 'throw a StorageNotAvailable exception, the storage is marked as unavailable in the UI and DAV'? Not any of this fancy FailedStorage things, which was only introduced for the backend and auth mechanism stuff.

**Comment by PVince81** [comment]
I'll wait for you guys to settle on the expected logic.

**Comment by icewind1991** [comment]
I don't see us making changes to the logic before 9.0, so I would suggest to merge this and backport it to stable9, for 9.1 we should look for a cleaner solution without requiring logic in files_external

**Comment by PVince81** [comment]
Hmm, for some reason the PROPFIND on the SMB now gives me this 500 response:
```
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Icewind\SMB\Exception\ConnectionRefusedException</s:exception>
  <s:message>smb://localhost/shared/</s:message>
</d:error>
```

Only happens on the "/smb" storage when the SMB server is off.

Shouldn't those exceptions somehow get translated to StorageNotAvailableException ?

**Comment by PVince81** [comment]
I had to manually set "available" to false, then enable the SMB server, then redo a PROPFIND.
The flag properly resets itself.

I'll make a separate ticket to make SMB behave properly.

:+1: 

@Xenopathic do you agree ?

**Comment by Xenopathic** [comment]
I don't like that the availability logic is now being spread across multiple places... but it's close to release day so sure. But please change it after 9.0 :smile:

**Comment by PVince81** [comment]
@icewind1991 can you make a ticket to change it in 9.1 ? And "change" also might require a discussion about the approach.

**Comment by PVince81** [comment]
Raised https://github.com/owncloud/core/issues/22826 to properly implement "storage not available" in SMB.

**Comment by PVince81** [comment]
@karlitschek backport for 9.0 to allow ext storages to properly recover from non-availability ?

**Comment by PVince81** [comment]
We'll need this very soon for RC2, can we count your comment as thumbs up @Xenopathic ?

**Comment by karlitschek** [comment]
please backport

**Comment by icewind1991** [comment]
stable9: https://github.com/owncloud/core/pull/22832

**Comment by icewind1991** [comment]
> can you make a ticket to change it in 9.1 ? And "change" also might require a discussion about the approach.

https://github.com/owncloud/core/issues/22833

**Comment by Xenopathic** [comment]
:+1:

**Comment by PVince81** [comment]
Unrelated fail in the new CI job:
```
07:26:47 PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 32 bytes) in /ssd/jenkins/workspace/server-master-linux-php5.4-ci@4/database/sqlite/label/SLAVE/apps/files_sharing/tests/middleware/sharingcheckmiddleware.php on line 150
```

**Pull request merged**